### DISCUSSION TITLE
Angular DevTools browser extension in Cypress browsers

### DISCUSSION THREAD

**Discussion by GrumpyMeow** [cypress-io/cypress#25343] [Questions and Help]
During development of our Cypress test-scripts it's sometimes necessary to add "data-testid" attributes to components. For our testers it's quite a chore to figure out which Angular-component they then need to modify. The "[Angular DevTools](https://angular.io/guide/devtools)" provides functionality to trace things back to Angular components. 

The ideal situation would be to have these extensions loaded in the Cypress-browsers. Unfortunately the extension and Cypress are not working together nicely. I'd managed to get the extension to load, but it displays an error about that the "Application is not started". 

Is there any thing else which needs to be configured, to also get it working in Cypress?

**Comment by portyu9** [comment]
Hi @GrumpyMeow 

This looks like a limitation that existed in Angular DevTools at the time this question was posted rather than additional Cypress configuration being required.

The application under test in the Cypress runner is embedded in an iframe. Angular later explicitly added **iframe support to Angular DevTools**, allowing Angular applications embedded in an iframe to be debugged and profiled.

So with current tooling I would first:

1. Update Angular DevTools to a current version.
2. Make sure the Angular application is a **development build**. Angular DevTools cannot inspect production-optimized Angular builds.
3. In Angular DevTools, make sure the Cypress AUT frame is the frame being inspected.

If you are loading Angular DevTools programmatically through Cypress's `before:browser:launch` hook, there is also a newer Chrome limitation to be aware of: standard Chrome 137+ no longer supports loading extensions that way. Use **Chrome for Testing** or **Chromium** instead.

So there shouldn't be a Cypress-specific flag needed to make Angular DevTools communicate with the AUT anymore. If a current Angular DevTools version still says that the application is not detected, I would next verify that the app is really running a development build and that DevTools is connected to the AUT iframe rather than the Cypress runner frame.

Please mark this answer as accepted if it helped, thank you.
